File: Coding-Interviews/04-二维数组中的查找.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    bool find(vector<int> const& x, int target)
    {
        if (x.empty())
            return false;
        int left = 0, right = x.size() - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (x[mid] == target)
                return true;
            if (x[mid] < target)
                left = mid + 1;
            else
                right = mid - 1;
        }
        return false;
    }
    bool findNumberIn2DArray(vector<vector<int>>& matrix, int target)
    {

        for (auto const& i : matrix) {
            bool is_found = find(i, target);
            if (is_found)
                return true;
        }
        return false;
    }
};

File: Coding-Interviews/04-二维数组中的查找.py
import bisect


# class Solution:
#     def findNumberIn2DArray(self, matrix: list[list[int]], target: int) -> bool:
#         for row in matrix:
#             i = 0
#             j = len(row)-1
#             # [i,j]
#             while i <= j:
#                 # j - i ==1时, mid = j
#                 mid = i + (j - i + 1)//2
#                 if row[mid] == target:
#                     return True
#                 elif row[mid] < target:
#                     i = mid + 1
#                 else:
#                     j = mid - 1
#         return False
# Z字查找, 将矩阵旋转45度就明白了, 从右上角开始作为mid, 每次消除一行或一列
class Solution:
    def findNumberIn2DArray(self, matrix: list[list[int]], target: int) -> bool:
        if matrix == []:
            return False
        x = 0
        y = len(matrix[0]) - 1
        while x < len(matrix) and y >= 0:
            if matrix[x][y] == target:
                return True
            elif matrix[x][y] < target:
                x += 1
            else:
                y -= 1
        return False


if __name__ == "__main__":
    a = [1, 2, 3, 4]
    idx = bisect.bisect_left(a, -1)
    print(idx)
